# A Nascent Peptide Code for Translational Control of mRNA Stability in Human Cells <!-- omit in toc -->

**Phil Burke**<sup>1,2</sup>, **Heungwon Park**<sup>1</sup>, **Arvind Rasi Subramaniam**<sup>1,†</sup>

<sup>1</sup> Basic Sciences Division and Computational Biology Section of the
Public Health Sciences Division, Fred Hutchinson Cancer Research Center, Seattle
<br/>
<sup>2</sup> Microbiology Graduate Program, University of Washington, Seattle <br/>

<sup>†</sup>Corresponding author: A.R.S: <rasi@fredhutch.org>

bioRxiv 2021.12.01.470782; doi: https://doi.org/10.1101/2021.12.01.470782

**Contents**

- [Abstract](#abstract)

## Abstract

Stability of eukaryotic mRNAs is associated with their codon, amino acid, and GC content.
Yet, coding sequence motifs that predictably alter mRNA stability in human cells remain poorly defined. 
Here, we develop a massively parallel assay to measure mRNA effects of thousands of synthetic and endogenous coding sequence motifs in human cells. 
We identify several families of simple dipeptide repeats whose translation triggers mRNA destabilization.
Rather than individual amino acids, specific combinations of bulky and positively charged amino acids are critical for the destabilizing effects of dipeptide repeats.
Remarkably, dipeptide sequences that form extended β strands *in silico* and *in vitro* drive ribosome stalling and mRNA instability *in vivo*.
The resulting nascent peptide code underlies ribosome stalling and mRNA-destabilizing effects of hundreds of endogenous peptide sequences in the human proteome.
Our work suggests an intrinsic role for the ribosome as a selectivity filter against the synthesis of bulky and aggregation-prone peptides.