---
title: Find barcodes for each insert
author: Arvind Rasi Subramaniam
date: 29 Oct 2020
---

**Edit this Rscript only in the accompanying .Rmd file with same name and
export by running the last cell in the .Rmd file.**

```{r setup, include=FALSE, purl=F}
knitr::opts_chunk$set(echo = TRUE, warning=F, message=F)
```

# Load libraries and define analysis-specific parameters
```{r}
args <- commandArgs(trailingOnly = T)
input_file <- args[1]
output_file <- args[2]
print(input_file)

library(Biostrings)
library(GenomicAlignments)
library(plyranges)
library(tidyverse)
```

# Read alignments
```{r}
bamfile <- BamFile(input_file)
# extract the number of mismatches and total edits
param <- ScanBamParam(
  # what = scanBamWhat(), 
  what = c('qname', 'flag'),
  # no secondary alignments
  flag = scanBamFlag(isSecondaryAlignment = F),
  # extract number of mismatches
  tag = c("XM"), 
  # limit number of mismatches to 2, 
  # there are typically <5% with >2 mismatches, but check this for each sample
  tagFilter = list("XM" = c(0, 1, 2)),
  # include only snps; exclude indels
  simpleCigar = T,
  # skip reads that have >1% chance of mapping to another ref
  mapqFilter = 20
)
alns <- readGAlignments(bamfile, param = param) 
```

# Write read-ref pairs to output
```{r}
alns %>% 
  GRanges() %>% 
  # The alignment has to start at the first nt of the reference
  filter(start(.) == 1) %>% 
  as_tibble() %>% 
  select(seqnames, qname, width, XM) %>% 
  rename(ref = seqnames, n_mismatch = XM) %>% 
  write_tsv(output_file)
```


```{r, purl=F}
knitr::purl("filter_alignments.Rmd", documentation = 2)
```