---
title: Filter barcodes to remove ones aligning to multiple inserts
author: Arvind Rasi Subramaniam
date: Oct 29, 2020
---

**Edit this Rscript only in the accompanying .Rmd file with same name and
export by running the last cell in the .Rmd file.**

```{r setup, include=FALSE, purl=F}
knitr::opts_chunk$set(echo = TRUE, warning=F, message=F)
```

# Load libraries and define analysis-specific parameters
```{r}
args <- commandArgs(trailingOnly = T)
barcode_alignment_file <- args[1]
barcode_insert_file <- args[2]
barcode_fasta_file <- args[3]
output_file <- args[4]
# barcode_alignment_file <- '../data/ref_vs_ref_alignments/didi_2_grna/alignment.bam'
# barcode_fasta_file <- '../data/ref_vs_ref_alignments/didi_2_grna/reference.fasta'
# barcode_insert_file <- '../data/insert_barcode_counts/didi_2_grna.tsv.gz'
# output_file <- '../data/filtered_barcodes/didi_2_grna.tsv.gz'

library(Biostrings)
library(GenomicAlignments)
library(plyranges)
library(tidyverse)
```

# Read barcode numbers
```{r}
barcode_nums <- readDNAStringSet(barcode_fasta_file) %>% 
  as.data.frame() %>% 
  rownames_to_column() %>% 
  setNames(c("barcode_num", "barcode_2")) %>% 
  as_tibble() %>% 
  print()
```

# Read insert-barcode pair counts 
```{r}
insert_barcodes <- read_tsv(barcode_insert_file) %>% 
  bind_cols(barcode_nums) %>% 
  print()
```


# Read barcode vs barcode alignments
```{r}
bamfile <- BamFile(barcode_alignment_file)
# extract the number of mismatches and total edits
param <- ScanBamParam(
  # what = scanBamWhat(),
  what = c("qname", "flag"),
  # extract number of mismatches
  tag = c("XM"), 
  # include only snps; exclude indels
  simpleCigar = T
)
alns <- readGAlignments(bamfile, param = param) %>% 
  as_tibble() %>% 
  rename(rname = seqnames) %>% 
  select(rname, qname, flag, XM) %>% 
  print()
```

# Find barcodes that are linked to distinct insert or might be sequencing errors

```{r}
exclude <- alns %>% 
  filter(rname != qname) %>% 
  left_join(select(insert_barcodes, insert_num, barcode_num, read_count), by = c("rname" = "barcode_num")) %>%
  rename(rinsert = insert_num, rcount = read_count) %>% 
  right_join(select(insert_barcodes, insert_num, barcode_num, read_count), by = c("qname" = "barcode_num")) %>%
  rename(qinsert = insert_num, qcount = read_count) %>% 
  # this exludes:
  # 1. barcodes that map to two distinct inserts
  # 2. barcodes that got lower count than another homologus barcode with same insert
  filter(!(qinsert == rinsert & qcount > rcount)) %>% 
  distinct(qname) %>% 
  print()
```

# Write barcodes that do not clash to output
```{r}
insert_barcodes %>% 
  anti_join(select(exclude, qname), by = c("barcode_num" = "qname")) %>% 
  select(insert_num, barcode_num, barcode, read_count) %>% 
  arrange(-read_count) %>% 
  mutate(barcode_num = 1:n()) %>% 
  write_tsv(output_file)
```


```{r, purl=F}
knitr::purl("filter_barcodes.Rmd", documentation = 2)
```