---
title: "Plot alignment statistics of barcode library"
author: "`r Sys.info()[['user']]`"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  github_document:
    toc: 2
---

```{r setup, include=FALSE, purl=F}
knitr::opts_chunk$set(echo = TRUE, warning=F, message=F)
```

# Load libraries and define analysis-specific parameters
```{r}
library(Biostrings)
library(plyranges)
library(tidyverse)
library(rasilabRtemplates)
```

# Read alignment log
```{r}
align_log <- list.files("../data/alignments/", pattern = "*.log", full.names = T) %>% 
  enframe("sno", "file") %>% 
  filter(str_detect(file, "endo12k")) %>% 
  mutate(log = map(file, readLines)) %>% 
  mutate(sample_name = str_extract(file, "[^/]+(?=.bowtie2align.log)")) %>% 
  select(-sno, -file) %>% 
  unnest() %>% 
  print()
```

# How many total reads and reads with expected barcodes?
```{r,fig.width=6, fig.height=3}
plot_data <- align_log %>% 
  mutate(total_all = str_extract(log, "^[:digit:]+(?= reads;)")) %>% 
  mutate(aln_unique = str_extract(log, "[:digit:]+(?=.+exactly)")) %>% 
  mutate(aln_multi = str_extract(log, "[:digit:]+(?=.+>1)")) %>% 
  select(-log) %>% 
  pivot_longer(-sample_name, names_to = 'type', values_to = 'count') %>% 
  filter(!is.na(count)) %>% 
  type_convert() %>% 
  separate(type, c("read", "type")) %>% 
  print()

plot_data %>% 
  ggplot(aes(x = fct_rev(read), y = count, fill = type)) +
  facet_wrap(~sample_name, ncol = 2, scales = "free_x") +
  geom_col() +
  ggthemes::scale_fill_colorblind() +
  theme_bw() +
  labs(x = "", y = "Read count",
       title = "Alignment statistics for barcode mRNA sequencing",
       subtitle = "Alignment against insert-linked barcodes")

ggsave("../figures/20210703_alignment_statistics.pdf")
```

# Get insert-barcode linkages with counts

```{r}
insert_barcode_linkages <- list.files("../data/filtered_barcodes//", 
                                      pattern = ".tsv.gz$", full.names = T) %>% 
  enframe("sno", "file") %>% 
  mutate(counts = map(file, read_tsv)) %>% 
  mutate(sample_name = str_extract(file, "[^/]+(?=.tsv.gz)")) %>% 
  select(-sno, -file) %>% 
  unnest() %>% 
  select(-barcode_1) %>%
  print()
```

# Get sample annotations
```{r}
sample_annotations <- read_csv("../annotations/sample_annotations.csv", comment = "#") %>% 
  print()
```

# Plot CDF of the number of reads per barcode
```{r,fig.width=6, fig.height=3}
plot_data <- insert_barcode_linkages %>% 
  group_by(sample_name) %>% 
  arrange(-read_count) %>% 
  mutate(xaxis = 1:(dplyr::n())) %>% 
  ungroup() %>% 
  print()

plot_data %>% 
  ggplot(aes(x = xaxis, y = read_count, color = sample_name)) +
  geom_line(aes(group = sample_name)) +
  theme_bw() +
  scale_y_log10() +
  labs(x = "Barcode number", y = "Number of reads",
       title = "Reads per barcode in each library") +
  scale_color_manual(values = cbPalette[2:3]) +
  NULL

ggsave("../figures/20210703_cdf_reads_per_barcode.pdf")
```

# Plot CDF of the number of reads per insert, summed over barcodes
```{r,fig.width=4, fig.height=3}
plot_data <- insert_barcode_linkages %>% 
  group_by(sample_name, insert_num) %>% 
  summarize(read_count = sum(read_count)) %>% 
  ungroup() %>% 
  group_by(sample_name) %>% 
  arrange(-read_count) %>% 
  mutate(xaxis = 1:(dplyr::n())) %>% 
  ungroup() %>% 
  print()

plot_data %>% 
  ggplot(aes(x = xaxis, y = read_count, color = sample_name)) +
  geom_line(aes(group = sample_name)) +
  theme_bw() +
  theme(legend.position = "top") +
  scale_y_log10() +
  labs(x = "Insert number", y = "Number of reads",
       title = "Reads per insert in each library") +
  scale_color_manual(values = cbPalette[2:3]) +
  NULL

ggsave("../figures/20210703_cdf_reads_per_barcode.pdf")
```